### DISCUSSION TITLE
[ws] Add '_socket:Socket' to WebSocket

### DISCUSSION THREAD

**Discussion by AaronNGray** [DefinitelyTyped/DefinitelyTyped#70452] [Issues with a @types package]
In @tyeps/ws add `_socket:Socket` to `WebSocket`

https://www.npmjs.com/package/ws has an instance variable called `_socket` of type `node:net` `Socket` which is useful for getting the remote client address amongst other purposes.
https://github.com/websockets/ws/blob/master/lib/websocket.js#L70

Example usage :-
```
        this.sockserver.clients.forEach((client:WebSocket) => {
            console.log(`distributing message: ${data}`)
            console.log(`client._socket.remoteAddress = `, client._socket.remoteAddress)
            client.send(data);
        });
```

There are also multiple othre instance variable that are not private that may also offer extra functionality to the programmer :- 

https://github.com/websockets/ws/blob/master/lib/websocket.js#L57C1-L71C1

[a comment by an automated account is not shown]

**Comment by AaronNGray** [comment]
I would give a pull request but there seem to be multiple lint errors on the @types/ws package. So would need advice on going forward. But I have it working locally with a patched `node_modules/@types/ws/index.d.ts` :-

https://gist.github.com/AaronNGray/f9d14d6391aa7f2c6f13018cd40d4670

**Comment by AaronNGray** [comment]
@einaros @lpinca

**Comment by reduckted** [comment]
`_socket` is a "private" field. If you want to access it, the first step would be changing `ws` to make it public. Then these types can be updated.

**Comment by AaronNGray** [comment]
Looks like it's pubic in the source, but just with an underscore. Will wait for authors to wade in, if not will make issue there.

**Comment by AaronNGray** [comment]
`ws.WebSocket` is in alignment with the HTML Living Standard which does not include socket access not even an getter or accessor.

So I will close the issue with the advice that the extra members maybe accessed including `_socket` using the following code :-
```
module ws {
    interface WebSocket {
        _socket: Socket;
    }
}
```
